### DISCUSSION TITLE
Retry even after task was success

### DISCUSSION THREAD

**Discussion by Raul824** [apache/airflow#36043] [Q&A]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

We have setup retry in case of failures and disconnects on individual task.
Some tasks are being reran post completion.

If a task is success it should not attempt to rerun it again.

At 8:05 the task was completed marked success by airflow.

[2023-12-04, 08:05:23 AEDT] {taskinstance.py:1350} INFO - Marking task as SUCCESS. dag_id=UDPPRDAU_ODS_KEY_SCD_SCF_06, task_id=SSOT_ODS_STG_GL_DTL_DATA_PAY_DTL.ODS_GL_DTL_DATA_PAY_DTL, execution_date=20231202T125900, start_date=20231203T210522, end_date=20231203T210523

But At 8:42 it again reran the same task and there are no clear events on this task in audit log.

*** Found local files:
***   * /opt/airflow/logs/dag_id=UDPPRDAU_ODS_KEY_SCD_SCF_06/run_id=manual__2023-12-02T12:59:00+00:00/task_id=SSOT_ODS_STG_GL_DTL_DATA_PAY_DTL.ODS_GL_DTL_DATA_PAY_DTL/attempt=3.log
[2023-12-04, 08:42:20 AEDT] {taskinstance.py:1097} INFO - Dependencies not met for <TaskInstance: UDPPRDAU_ODS_KEY_SCD_SCF_06.SSOT_ODS_STG_GL_DTL_DATA_PAY_DTL.ODS_GL_DTL_DATA_PAY_DTL manual__2023-12-02T12:59:00+00:00 [success]>, dependency 'Task Instance State' FAILED: Task is in the 'success' state.
[2023-12-04, 08:42:20 AEDT] {local_task_job_runner.py:154} INFO - Task is not able to be run
[2023-12-04, 09:45:29 AEDT] {taskinstance.py:1103} INFO - Dependencies all met for dep_context=non-requeueable deps ti=<TaskInstance: UDPPRDAU_ODS_KEY_SCD_SCF_06.SSOT_ODS_STG_GL_DTL_DATA_PAY_DTL.ODS_GL_DTL_DATA_PAY_DTL manual__2023-12-02T12:59:00+00:00 [queued]>
[2023-12-04, 09:45:29 AEDT] {taskinstance.py:1103} INFO - Dependencies all met for dep_context=requeueable deps ti=<TaskInstance: UDPPRDAU_ODS_KEY_SCD_SCF_06.SSOT_ODS_STG_GL_DTL_DATA_PAY_DTL.ODS_GL_DTL_DATA_PAY_DTL manual__2023-12-02T12:59:00+00:00 [queued]>
[2023-12-04, 09:45:29 AEDT] {taskinstance.py:1308} INFO - Starting attempt 3 of 501
[2023-12-04, 09:45:29 AEDT] {taskinstance.py:1327} INFO - Executing <Task(PythonOperator): SSOT_ODS_STG_GL_DTL_DATA_PAY_DTL.ODS_GL_DTL_DATA_PAY_DTL> on 2023-12-02 12:59:00+00:00
[2023-12-04, 09:45:29 AEDT] {standard_task_runner.py:57} INFO - Started process 1128 to run task

My understanding was that retries comes into picture only in case of failures, which working fine in most of the tasks but it randomly reruns the tasks even after success.

### What you think should happen instead

Once the task is success, airflow should not retry the task.

### How to reproduce

Set up_for_retry for all tasks and run 200 tasks at the same time.

### Operating System

Azure Kubernetes Services

### Versions of Apache Airflow Providers

Airflow 2.6.1

### Deployment

Official Apache Airflow Helm Chart

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by Raul824** [comment]
Below is the snippet of another such instance.
All tasks in a task group are sequential with trigger rule all_success.
max active run of dag is 1 and depends on past is true.

The dag is success even the task CMMS_DIGT_ELY is successful in logs but it went in up_for_retry status without any external interference. The next dag started running as soon as the previous one was success but task went in up_for_retry and due to depends on past as True it was stuck.

![image](https://github.com/apache/airflow/assets/51978647/c3919aea-7fae-4d9c-a0dd-2578025f09f1)

**Comment by eladkal** [comment]
Coverting to discussion as you didnt add reproduce example.
Please share reproduce example

**Comment by mkavakebi** [comment]
I have the same Problem. the first try is successfull, but still it starts the second try.
in another word it randomly runs two times and both are marked as success

**Reply by potiuk** [reply]
How about providing more details, logs and describing your deployment? Without it, it does not bring anyone closer to being able to diagnose it.

**Reply by Raul824** [reply]
Sorry, but I don't have any steps to replicate it yet. I would have built a workaround to handle the exceptions if I would have been able to replicate it.

So far the only thing I know is that even after success the postgres table was still showing the task as up_for_retry in task_instance. So my best guess would be that some issue in status update in postgres database.

**Comment by asm1213** [comment]
We periodically get the same problem on two installations with versions 2.8.3 and 2.9.2

**Reply by Raul824** [reply]
After no reply I checked the logs and found out that the postgres table was not updating due to some queries might be timing out or some connection issues. So I created a workaround where if dag state is success and task is in up for retry I mark them as success. 
I scheduled a dag on this logic to run hourly.

Below is the function which I created if you want to use it you can but it's more of a tactical solution. And in the logs you see which tasks are marked as success so far this is working fine for me and is not causing any issues.

Snippet from logs.

![image](https://github.com/user-attachments/assets/d0dec40a-5962-467c-9a67-fdff2f0eb827)


```
def upretry_fix():
    import subprocess
    sel_retry_qry = """select
ti.task_id
,ti.state
,ti.dag_id
,ti.run_id
,dr.state dag_state
from task_instance ti
inner join dag_run dr
on dr.run_id = ti.run_id
and dr.dag_id = ti.dag_id
and ti.state = 'up_for_retry'
and dr.state = 'success'
where ti.state = 'up_for_retry'
and dr.state = 'success'"""
    upretry_tasks = f"""psql $AIRFLOW__CORE__SQL_ALCHEMY_CONN -c "{sel_retry_qry}" """
    try:
        output = subprocess.run(upretry_tasks,stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True, text=True, check=True)
        print(output.stdout)
    except subprocess.CalledProcessError as e:
        error = "SHELL error" + str(output.stderr)
        print(error)
        raise Exception(error)
    print(output.stdout)
    try:
        task_list = output.stdout.split('-\n ')[1].split("\n(")[0].split('\n ')
        print("----Marking below tasks to success task_list")
        print(task_list)
    except IndexError:
        task_list = []
        print("No Tasks found which are stuck in up for retry status")

    if len(task_list) > 0:
        upd_retry_qry = """Update task_instance
set state = 'success'
from
(
select
ti.task_id
,ti.state
,ti.dag_id
,ti.run_id
,dr.state dag_state
from task_instance ti
inner join dag_run dr
on dr.run_id = ti.run_id
and dr.dag_id = ti.dag_id
and ti.state = 'up_for_retry'
and dr.state = 'success'
where ti.state = 'up_for_retry'
and dr.state = 'success'
) dt
where task_instance.task_id  = dt.task_id
and task_instance.dag_id = dt.dag_id
and task_instance.run_id = dt.run_id
and task_instance.state = 'up_for_retry'"""
        pause_dag_cmd = f""" psql $AIRFLOW__CORE__SQL_ALCHEMY_CONN -c "{upd_retry_qry}" """
        try:
            ps_output = subprocess.run(pause_dag_cmd,stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True, text=True, check=True)
            print(ps_output.stdout)
        except subprocess.CalledProcessError as e:
            error = "SHELL error" + str(ps_output.stderr)
            print(error)
            raise Exception(error)
    else:
        print("No Tasks found which are stuck in up for retry status")
```

**Reply by asm1213** [reply]
Thx!
